File: solutions/haskell/list-ops/1/src/ListOps.hs
module ListOps
  ( length
  , reverse
  , map
  , filter
  , foldr
  , foldl'
  , (++)
  , concat
  ) where

import Prelude hiding
  ( length, reverse, map, filter, foldr, (++), concat )

foldl' :: (b -> a -> b) -> b -> [a] -> b
foldl' _ z [] = z
foldl' f z (x:xs) =
  let z' = f z x
  in z' `seq` foldl' f z' xs

foldr :: (a -> b -> b) -> b -> [a] -> b
foldr _ z [] = z
foldr f z (x:xs) = f x (foldr f z xs)

length :: [a] -> Int
length = foldr (\_ a -> 1 + a) 0

reverse :: [a] -> [a]
reverse = foldl' (flip (:)) []

map :: (a -> b) -> [a] -> [b]
map _ [] = []
map f (x:xs) = (f x) : (map f xs)

filter :: (a -> Bool) -> [a] -> [a]
filter _ [] = []
filter f (x:xs)
  | f x = x : (filter f xs)
  | otherwise = filter f xs

(++) :: [a] -> [a] -> [a]
xs ++ [] = xs
xs ++ ys = foldr (:) ys xs

concat :: [[a]] -> [a]
concat = foldr (++) [] 

File: solutions/csharp/list-ops/1/ListOps.cs
public static class ListOps
{
    public static int Length<T>(List<T> input) =>
        input.Count;

    public static List<T> Reverse<T>(List<T> input) =>
        input.AsEnumerable().Reverse().ToList();

    public static List<U> Map<T, U>(List<T> input, Func<T, U> map) =>
        input.Select(map).ToList();

    public static List<T> Filter<T>(List<T> input, Func<T, bool> predicate) =>
        input.Where(predicate).ToList();

    public static U Foldl<T, U>(List<T> input, U start, Func<U, T, U> func) =>
        input.Aggregate(start, func);

    public static U Foldr<T, U>(List<T> input, U start, Func<T, U, U> func) =>
        Reverse(input).Aggregate(start, (el, acc) => func(acc, el));

    public static List<T> Concat<T>(List<List<T>> input) =>
        input.SelectMany(e => e).ToList();

    public static List<T> Append<T>(List<T> left, List<T> right) =>
        left.Concat(right).ToList();
}